hub-saml/src/main/java/uk/gov/ida/saml/core/validators/assertion/DuplicateAssertionValidatorImpl.java

```diff
@@ -0,0 +1,43 @@
+package uk.gov.ida.saml.core.validators.assertion;
+
+import com.google.inject.Inject;
+import org.joda.time.DateTime;
+import org.opensaml.saml.saml2.core.Assertion;
+import uk.gov.ida.saml.hub.exception.SamlValidationException;
+
+import java.util.concurrent.ConcurrentMap;
+
+import static uk.gov.ida.saml.core.errors.SamlTransformationErrorFactory.authnStatementAlreadyReceived;
+import static uk.gov.ida.saml.core.errors.SamlTransformationErrorFactory.duplicateMatchingDataset;
+
+public class DuplicateAssertionValidatorImpl implements DuplicateAssertionValidator {
+
+    private final ConcurrentMap<String, DateTime> duplicateIds;
+
+    @Inject
+    public DuplicateAssertionValidatorImpl(ConcurrentMap<String, DateTime> duplicateIds) {
+        this.duplicateIds = duplicateIds;
+    }
+
+    @Override
+    public void validateAuthnStatementAssertion(Assertion assertion) {
+        if (!valid(assertion)) throw new SamlValidationException(authnStatementAlreadyReceived(assertion.getID()));
+    }
+
+    @Override
+    public void validateMatchingDataSetAssertion(Assertion assertion, String responseIssuerId) {
+        if (!valid(assertion)) throw new SamlValidationException(duplicateMatchingDataset(assertion.getID(), responseIssuerId));
+    }
+
+    private boolean valid(Assertion assertion) {
+        if (isDuplicateNonExpired(assertion)) return false;
+
+        DateTime expire = assertion.getSubject().getSubjectConfirmations().get(0).getSubjectConfirmationData().getNotOnOrAfter();
```

Comment:
This is a very long line, what happens if any of the elements are null?

Reply:
I believe they are checked before this is called.